File: Notebooks/bayes_net_utils.R
bayes_net_predict <- function(rfile_fpath, sd_fpath, year, chla_prevSummer, colour_prevSummer, TP_prevSummer, wind_speed, rain, sigma) {
    #' Make predictions given the evidence provided based on the pre-fitted Bayesian network (saved as
    #' 'Vansjo_fitted_seasonal_GaussianBN_1981-2018.rds').
    #'
    #' Args:
    #'     rfile_fpath:       Str. Filepath to fitted BNLearn network object (.rds file)
    #'     sd_fpath:          Str. Filepath to csv containing standard deviation info from fitted BN
    #'     boundaries_list    Named list of thresholds to use when classifying. Must include labels 'TP', 'chla', 'colour', 'cyano' 
    #'     year:              Int. Year for prediction
    #'     chla_prevSummer:   Float. Chl-a measured from the previous summer (mg/l)
    #'     colour_prevSummer: Float. Colour measured from the previous summer (mg Pt/l)
    #'     TP_prevSummer:     Float. Total P measured from the previous summer (mg/l)
    #'     wind_speed:        Float. Predicted wind speed for season of interest (m/s)
    #'     rain:              Float. Predicted precipitation for season of interest (mm)
    #'     sigma:             Float. Standard deviation of the Box-cox transformed observations, for back-transforming predictions
    #'
    #' Returns:
    #'     R dataframe

    suppressMessages(library(tidyverse))
    suppressMessages(library(bnlearn))

    # Build dataframe from input data
    year <- c(year)
    chla_prevSummer <- c(chla_prevSummer)
    colour_prevSummer <- c(colour_prevSummer)
    TP_prevSummer <- c(TP_prevSummer)
    wind_speed <- c(wind_speed)
    rain <- c(rain)

    driving_data <- data.frame(chla_prevSummer, colour_prevSummer, TP_prevSummer, wind_speed, rain)
    row.names(driving_data) <- year

    # Convert any integer cols to numeric
    driving_data[1:ncol(driving_data)] = lapply(driving_data[1:ncol(driving_data)], as.numeric)

    # Load fitted Bayesian network and pre-calculated std. devs.
    fitted_BN = readRDS(rfile_fpath)

    sds = read.csv(file=sd_fpath, header=TRUE, sep=",")

    # Nodes to make predictions for. Must match nodes present in the fitted BN.
    # Add check that list is sorted alphabetically, as concatenation of final df assumes this
    nodes_to_predict = sort(c('chla','colour','cyano', 'TP'))

    # Just select values associated with nodes for prediction, and sort alphabetically
    sd_predictedNodes = filter(sds, node %in% nodes_to_predict)
    sd_predictedNodes = sd_predictedNodes[order(sd_predictedNodes$node),]

    set.seed(1)

    # Loop over nodes and derive expected values
    expectedValue_li = vector(mode = "list", length = 0)

    for (node in nodes_to_predict){
        pred = predict(fitted_BN,
                       data=driving_data,
                       node=node,
                       method='bayes-lw',
                       n=10000
                      )

        # If node is cyano, then remove the boxcox transformation before adding expected value to list
        if (node == "cyano") {

            # Without bias-adjustment, to estimate the median
            # pred = (pred*0.1 + 1)**(1/0.1) # 0.1 is lambda value chosen in transformation

            # With bias-adjusted back transformation, to estimate the mean
            pred = ((pred * 0.1 + 1) ** (1 / 0.1)) * (1 + ((sigma**2 * (1 - 0.1)) / (2 * (0.1 * pred + 1)**2)))
        }
        expectedValue_li[[node]] = pred # Update list with value for this node
    }

    # Sort alphabetically
    expectedValue_li = expectedValue_li[order(names(expectedValue_li))]

    # Classify

    # Thresholds to use in classification
    boundaries_list = list('TP' = 29.5,     # Middle of 'Moderate' class
                           'chla' = 20.0,   # M-P boundary. WFD boundaries: [10.5, 20.0]. Only 6 observed points under 10.5 so merge G & M
                           'colour' = 48.0, # 66th percentile (i.e. upper tercile). No management implications
                           'cyano' = 1.0    # M-P boundary is 2.0, but there were only 2 values in this class. Plenty above 2 tho
                           )

    # Sort alphabetically
    boundaries_list = boundaries_list[order(names(boundaries_list))]

    # Data for evidence, converted to named list
    evidence_li = as.list(driving_data)

    # Empty list to be populated with probability of being below boundary
    prob_li = vector(mode = "list", length = 0)

    # Loop over nodes and predict probabilities of being within various classes
    for (node in nodes_to_predict) {
        boundary = unlist(boundaries_list[node], use.names = FALSE)

        # If cyanomax, apply boxcox transformation with lambda=0.1 to boundary first
        if (node == 'cyano') {
            boundary = (boundary^0.1 - 1) / 0.1
        }

        # Horrible hack to make 'query' available in the global namespace. See
        #     https://stackoverflow.com/q/19260580/505698
        query <<- paste("(", node, ' < ', boundary, ")", sep = "")
        prob = cpquery(fitted_BN,
                       event = eval(parse(text = query)),
                       evidence = evidence_li,
                       method = 'lw'
                      )

        # Round to 2 d.p. Below this, cpquery returns variable results over diff calls
        # Even with rounding, still get some variability in results
        prob = round(prob, digits = 2)

        prob_li[[node]] = prob
    }

    # Sort alphabetically
    prob_li = prob_li[order(names(prob_li))] # Sort alphabetically

    # Build dataframe
    prob_df = data.frame(node = nodes_to_predict,
                         threshold = unlist(boundaries_list, use.names = FALSE),
                         prob_below_threshold = unlist(prob_li, use.names = FALSE),
                         prob_above_threshold = 1 - unlist(prob_li, use.names = FALSE),
                         expected_value = signif(unlist(expectedValue_li, use.names = FALSE), 3), #Round to 3 s.f
                         st_dev = signif(sd_predictedNodes['sd'], 3)                           # Round to 3 s.f
                        )

    return(prob_df)
}

bayes_net_predict_nomet <- function(rfile_fpath, sd_fpath,
                              year, chla_prevSummer, colour_prevSummer, TP_prevSummer, sigma) {
    #' Make predictions given the evidence provided based on the pre-fitted Bayesian network with no met nodes (saved as 
    #' 'Vansjo_fitted_seasonal_GaussianBN_1981-2018.rds'). 
    #' 
    #' Args:
    #'     rfile_fpath:       Str. Filepath to fitted BNLearn network object (.rds file)
    #'     sd_fpath:          Str. Filepath to csv containing standard deviation info from fitted BN
    #'     year:              Int. Year for prediction
    #'     chla_prevSummer:   Float. Chl-a measured from the previous summer (mg/l)
    #'     colour_prevSummer: Float. Colour measured from the previous summer (mg Pt/l)
    #'     TP_prevSummer:     Float. Total P measured from the previous summer (mg/l)
    #'     sigma:             Float. Standard deviation of the Box-cox transformed observations, for back-transforming predictions
    #' 
    #' Returns:
    #'     R dataframe

    suppressMessages(library(tidyverse))
    suppressMessages(library(bnlearn))

    # Build dataframe from input data
    year <- c(year)
    chla_prevSummer <- c(chla_prevSummer)
    colour_prevSummer <- c(colour_prevSummer)
    TP_prevSummer <- c(TP_prevSummer)

    driving_data <- data.frame(chla_prevSummer, colour_prevSummer, TP_prevSummer)
    row.names(driving_data) <- year

    # Convert any integer cols to numeric
    driving_data[1:ncol(driving_data)] = lapply(driving_data[1:ncol(driving_data)], as.numeric)

    # Load fitted Bayesian network and pre-calculated std. devs.
    fitted_BN = readRDS(rfile_fpath)
    sds = read.csv(file = sd_fpath, header = TRUE, sep = ",")

    # Nodes to make predictions for. Must match nodes present in the fitted BN.
    # Add check that list is sorted alphabetically, as concatenation of final df assumes this
    nodes_to_predict = sort(c('chla', 'colour', 'cyano', 'TP'))

    # Just select values associated with nodes for prediction, and sort alphabetically
    sd_predictedNodes = filter(sds, node %in% nodes_to_predict)
    sd_predictedNodes = sd_predictedNodes[order(sd_predictedNodes$node),]

    set.seed(1)

    # Loop over nodes and derive expected values
    expectedValue_li = vector(mode = "list", length = 0)

    for (node in nodes_to_predict) {
        pred = predict(fitted_BN,
                       data = driving_data,
                       node = node,
                       method = 'bayes-lw',
                       n = 10000
                      )

        # If node is cyano, then remove the boxcox transformation before adding expected value to list
        if (node == "cyano")  {
            # sigma = subset(sd_predictedNodes, node == "cyano")$sd

            # Without bias-adjustment, to estimate the median
            # pred = (pred*0.1 + 1)**(1/0.1) # 0.1 is lambda value chosen in transformation

            # With bias-adjusted back transformation, to estimate the mean
            pred = (pred*0.1 + 1)**(1/0.1)*(1 + ((sigma**2 * (1-0.1))/(2*(0.1*pred+1)**2)))
        }
        expectedValue_li[[node]] = pred # Update list with value for this node
    }

    # Sort alphabetically
    expectedValue_li = expectedValue_li[order(names(expectedValue_li))]

    # Thresholds to use in classification
    boundaries_list = list('TP' = 29.5,     # Middle of 'Moderate' class
                           'chla' = 20.0,   # M-P boundary. WFD boundaries: [10.5, 20.0]. Only 6 observed points under 10.5 so merge G & M
                           'colour' = 48.0, # 66th percentile (i.e. upper tercile). No management implications
                           'cyano' = 1.0    # M-P boundary is 2.0, but there were only 2 values in this class. Plenty above 2 tho
                           )

    # Sort alphabetically
    boundaries_list = boundaries_list[order(names(boundaries_list))]

    # Data for evidence, converted to named list
    evidence_li = as.list(driving_data)

    # Empty list to be populated with probability of being below boundary
    prob_li = vector(mode = "list", length = 0)

    # Loop over nodes and predict probabilities of being within various classes
    for (node in nodes_to_predict){
        boundary = unlist(boundaries_list[node], use.names=FALSE)

        # If cyanomax, apply boxcox transformation with lambda=0.1 to boundary first
        if (node=='cyano') {
            boundary = (boundary^0.1 - 1) / 0.1
        }

        # Horrible hack to make 'query' available in the global namespace. See
        #     https://stackoverflow.com/q/19260580/505698
        query <<- paste("(", node, ' < ', boundary, ")", sep = "")
        prob = cpquery(fitted_BN,
                       event = eval(parse(text = query)),
                       evidence = evidence_li,
                       method = 'lw'
                      )

        # Round to 2 d.p. Below this, cpquery returns variable results over diff calls
        # Even with rounding, still get some variability in results
        prob = round(prob, digits = 2)

        prob_li[[node]] = prob
    }

    # Sort alphabetically
    prob_li = prob_li[order(names(prob_li))] # Sort alphabetically

    # Build dataframe
    prob_df = data.frame(node = nodes_to_predict,
                         threshold = unlist(boundaries_list, use.names = FALSE),
                         prob_below_threshold = unlist(prob_li, use.names = FALSE),
                         prob_above_threshold = 1 - unlist(prob_li, use.names = FALSE),
                         expected_value = signif(unlist(expectedValue_li, use.names = FALSE), 3), #Round to 3 s.f
                         st_dev = signif(sd_predictedNodes['sd'], 3)                           # Round to 3 s.f
                        )

    return(prob_df)
}


bayes_net_predict_operational <- function(rfile_fpath,
                                          year, chla_prevSummer, colour_prevSummer, TP_prevSummer, sigma) {
    #' Make predictions given the evidence provided based on the pre-fitted Bayesian network (saved as 
    #' 'Vansjo_fitted_seasonal_GaussianBN_1981-2019.rds'. Drop met nodes for operational forecast,
    #' and only forecast for TP, colour and cyano (not chla)). Remove standard deviation info.
    #' 
    #' Args:
    #'     rfile_fpath:       Str. Filepath to fitted BNLearn network object (.rds file)
    #'     year:              Int. Year for prediction
    #'     chla_prevSummer:   Float. Chl-a measured from the previous summer (mg/l)
    #'     colour_prevSummer: Float. Colour measured from the previous summer (mg Pt/l)
    #'     TP_prevSummer:     Float. Total P measured from the previous summer (mg/l)
    #'     sigma:             Float. Standard deviation of the Box-cox transformed observations, for back-transforming predictions
    #' 
    #' Returns:
    #'     R dataframe

    suppressMessages(library(tidyverse))
    suppressMessages(library(bnlearn))

    # Build dataframe from input data
    year <- c(year)
    chla_prevSummer <- c(chla_prevSummer)
    colour_prevSummer <- c(colour_prevSummer)
    TP_prevSummer <- c(TP_prevSummer)

    driving_data <- data.frame(chla_prevSummer, colour_prevSummer, TP_prevSummer)
    row.names(driving_data) <- year

    # Convert any integer cols to numeric
    driving_data[1:ncol(driving_data)] = lapply(driving_data[1:ncol(driving_data)], as.numeric)

    # Load fitted Bayesian network and pre-calculated std. devs
    fitted_BN = readRDS(rfile_fpath)

    # Nodes to make predictions for. Must match nodes present in the fitted BN.
    # Add check that list is sorted alphabetically, as concatenation of final df assumes this
    nodes_to_predict = sort(c('colour','cyano', 'TP'))

    # Nodes to use when making predictions
    nodes_to_use = list('TP' = c('TP_prevSummer', 'chla_prevSummer'),
                        'colour' = c('colour_prevSummer'),
                        'cyano' = c("chla_prevSummer", "TP_prevSummer", "colour_prevSummer"))

    set.seed(1)

    # Loop over nodes and derive expected values
    expectedValue_li = vector(mode = "list", length = 0)

    for (node in nodes_to_predict){
        pred = predict(fitted_BN,
                       data=driving_data,
                       node=node,
                       method='bayes-lw',
                       from = unlist(nodes_to_use[node], use.names=FALSE),
                       n=10000
                      )

        # If node is cyano, then remove the boxcox transformation before adding expected value to list
        if (node == "cyano") {

            # Without bias-adjustment, to estimate the median
            # pred = (pred*0.1 + 1)**(1/0.1) # 0.1 is lambda value chosen in transformation

            # With bias-adjusted back transformation, to estimate the mean
            pred = (pred*0.1 + 1)**(1/0.1)*(1 + ((sigma**2 * (1-0.1))/(2*(0.1*pred+1)**2)))
        }
        expectedValue_li[[node]] = pred # Update list with value for this node
    }

    # Sort alphabetically
    expectedValue_li = expectedValue_li[order(names(expectedValue_li))]

    # Thresholds to use in classification
    boundaries_list = list('TP' = 29.5,     # Middle of 'Moderate' class
                           'colour' = 48.0, # 66th percentile (i.e. upper tercile). No management implications
                           'cyano' = 1.0    # M-P boundary is 2.0, but there were only 2 values in this class. Plenty above 2 tho
                           )

    # Sort alphabetically
    boundaries_list = boundaries_list[order(names(boundaries_list))]

    # Data for evidence, converted to named list
    evidence_li = as.list(driving_data)

    # Empty list to be populated with probability of being below boundary
    prob_li = vector(mode = "list", length = 0)

    # Loop over nodes and predict probabilities of being within various classes
    for (node in nodes_to_predict){
        boundary = unlist(boundaries_list[node], use.names=FALSE)

        # If cyanomax, apply boxcox transformation with lambda=0.1 to boundary first
        if (node == 'cyano') {
            boundary = (boundary^0.1 - 1) / 0.1
        }

        # Horrible hack to make 'query' available in the global namespace. See
        #     https://stackoverflow.com/q/19260580/505698
        query <<- paste("(", node, ' < ', boundary, ")", sep="")
        prob = cpquery(fitted_BN,
                       event = eval(parse(text = query)),
                       evidence=evidence_li,
                       method='lw'
                      )

        # Round to 2 d.p. Below this, cpquery returns variable results over diff calls
        # Even with rounding, still get some variability in results
        prob = round(prob,digits=2)

        prob_li[[node]] = prob
    }

    # Sort alphabetically
    prob_li = prob_li[order(names(prob_li))] # Sort alphabetically

    # Build dataframe
    prob_df = data.frame(node=nodes_to_predict,
                         threshold = unlist(boundaries_list, use.names=FALSE),
                         prob_below_threshold = unlist(prob_li, use.names=FALSE),
                         prob_above_threshold = 1-unlist(prob_li, use.names=FALSE),
                         expected_value = signif(unlist(expectedValue_li, use.names=FALSE),3) #Round to 3 s.f
                        )

    return(prob_df)
}

File: Notebooks/bayes_net_utils.py
def bayes_net_predict(rfile_fpath, sd_fpath, year, chla_prev_summer, colour_prev_summer,
                      tp_prev_summer, wind_speed, rain, sigma):
    """ Make predictions given the evidence provided based on a pre-fitted
        Bayesian network. This function is just a thin "wrapper" around the
        R function named 'bayes_net_predict' in 'bayes_net_utils.R'.
        See also bayes_net_predict_operational.

        NOTE: 'bayes_net_utils.R' must be in the same folder as this file.

    Args:
        rfile_fpath:       Str. Filepath to fitted BNLearn network object (.rds file)
        sd_fpath:          Str. Filepath to csv containing standard deviation info from fitted BN
        year:              Int. Year for prediction
        chla_prevSummer:   Float. Chl-a measured from the previous summer  (mg/l)
        colour_prevSummer: Float. Colour measured from the previous summer (mg Pt/l)
        TP_prevSummer:     Float. Total P measured from the previous summer (mg/l)
        wind_speed:        Float. Predicted wind speed for season of interest (m/s)
        rain:              Float. Predicted precipitation for season of interest (mm)
        sigma:             Float. Standard deviation of the box cox transformed observations

    Returns:
        Dataframe with columns 'year', 'node', 'threshold', 'prob_below_threshold',
       'prob_above_threshold', 'expected_value', 'sd' (standard deviation)
    """
    import pandas as pd
    import rpy2.robjects as ro
    from rpy2.robjects.packages import importr
    from rpy2.robjects import pandas2ri
    from rpy2.robjects.conversion import localconverter

    # Load R script
    ro.r.source('bayes_net_utils.R')

    # Call R function with user-specified evidence
    res = ro.r['bayes_net_predict'](rfile_fpath, sd_fpath, year,
                                    chla_prev_summer, colour_prev_summer,
                                    tp_prev_summer, wind_speed, rain, sigma)

    # Convert back to Pandas df
    with localconverter(ro.default_converter + pandas2ri.converter):
        df = ro.conversion.rpy2py(res)

    # Add 'year' to results as unique identifier
    df['year'] = int(year)
    df.reset_index(drop=True, inplace=True)

    # Add predicted WFD class
    df['WFD_class'] = df[['threshold', 'expected_value']].apply(lambda x: discretize([x.threshold], x.expected_value), axis=1)

    return df


def bayes_net_predict_nomet(rfile_fpath, sd_fpath, year, chla_prev_summer,
                            colour_prev_summer, tp_prev_summer, sigma):
    """ Make predictions given the evidence provided based on a pre-fitted
        Bayesian network with no met nodes
        This function is just a thin "wrapper" around the R function named
        'bayes_net_predict_nomet' in 'bayes_net_utils.R'.
        See also bayes_net_predict_operational.

        NOTE: 'bayes_net_utils.R' must be in the same folder as this file.

    Args:
        rfile_fpath:       Str. Filepath to fitted BNLearn network object (.rds file)
        sd_fpath:          Str. Filepath to csv with standard deviations from fitted BN
        year:              Int. Year for prediction
        chla_prevSummer:   Float. Chl-a measured from the previous summer  (mg/l)
        colour_prevSummer: Float. Colour measured from the previous summer (mg Pt/l)
        TP_prevSummer:     Float. Total P measured from the previous summer (mg/l)
        sigma:             Float. Standard deviation of the box cox transformed observations

    Returns:
        Dataframe with columns 'year', 'node', 'threshold', 'prob_below_threshold',
       'prob_above_threshold', 'expected_value', 'sd' (standard deviation)
    """
    import pandas as pd
    import rpy2.robjects as ro
    from rpy2.robjects.packages import importr
    from rpy2.robjects import pandas2ri
    from rpy2.robjects.conversion import localconverter

    # Load R script
    ro.r.source('bayes_net_utils.R')

    # Call R function with user-specified evidence
    res = ro.r['bayes_net_predict_nomet'](rfile_fpath, sd_fpath, year,
                                          chla_prev_summer, colour_prev_summer,
                                          tp_prev_summer, sigma)

    # Convert back to Pandas df
    with localconverter(ro.default_converter + pandas2ri.converter):
        df = ro.conversion.rpy2py(res)

    # Add 'year' to results as unique identifier
    df['year'] = int(year)
    df.reset_index(drop=True, inplace=True)

    # Add predicted WFD class
    df['WFD_class'] = df[['threshold', 'expected_value']].apply(lambda x: discretize([x.threshold], x.expected_value), axis=1)

    return df


def bayes_net_predict_operational(rfile_fpath, year, chla_prev_summer, colour_prev_summer,
                                  tp_prev_summer, sigma):
    """ Make predictions given the evidence provided based on the pre-fitted
        Bayesian network.
        This function is just a thin "wrapper" around the R function named
        'bayes_net_predict_operational' in 'bayes_net_utils.R'.

        Very similar to bayes_net_predict, but:
        - Drop met nodes for operational forecast
        - Only forecast for TP, colour and cyano (not chla)).
        - Remove standard deviations which were in bayes_net_predict function

        NOTE: 'bayes_net_utils.R' must be in the same folder as this file.

    Args:
        rfile_fpath:       Str. Filepath to fitted BNLearn network object (.rds file)
        year:              Int. Year for prediction
        chla_prevSummer:   Float. Chl-a measured from the previous summer  (mg/l)
        colour_prevSummer: Float. Colour measured from the previous summer (mg Pt/l)
        TP_prevSummer:     Float. Total P measured from the previous summer (mg/l)
        sigma:             Float. Standard deviation of the box cox transformed observations

    Returns:
        Dataframe with columns 'year', 'node', 'threshold', 'prob_below_threshold',
       'prob_above_threshold', 'expected_value'
    """
    import pandas as pd
    import rpy2.robjects as ro
    from rpy2.robjects.packages import importr
    from rpy2.robjects import pandas2ri
    from rpy2.robjects.conversion import localconverter

    # Load R script
    ro.r.source('bayes_net_utils.R')

    # Call R function with user-specified evidence
    res = ro.r['bayes_net_predict_operational'](rfile_fpath, year, chla_prev_summer,
                                                colour_prev_summer,
                                                tp_prev_summer, sigma)

    # Convert back to Pandas df
    with localconverter(ro.default_converter + pandas2ri.converter):
        df = ro.conversion.rpy2py(res)

    # Add 'year' to results as unique identifier
    df['year'] = int(year)
    df.reset_index(drop=True, inplace=True)

    # Add predicted WFD class
    df['WFD_class'] = df[['threshold','expected_value']].apply(lambda x: discretize([x.threshold], x.expected_value), axis=1)

    return df


def classification_error(obs, pred):
    """
    Calculate classification error, the proportion of time the model predicted
    the class correctly

    Input:
        obs: series of observed classes, numeric format
        pred: series of predicted classes (aligned to obs), numeric format
    Output:
        Classification error (float)
    """
    import pandas as pd
    import numpy as np

    assert len(obs) == len(pred), "observed and predicted series have different lengths"

    # Were observed and predictions in the same class? 'right' col is a boolean,
    # 1=Yes the same, 0=no different
    right = np.where((obs == pred), 1, 0)

    classification_error = 1 - (right.sum() / len(right))

    return classification_error


def daily_to_summer_season(daily_df, agg_method_dict, replace_zero_with_NaN_li=[],
                           return_just_summer=True):
    """
    Take a dataframe with daily frequency data, and aggregate it to seasonal
    (6 monthly), just picking results for the summer (May-Oct) season.
    Input:
    - dataframe of daily data
    - dictionary of aggregation method functions
      (e.g. np.nansum, with a key for each column in daily_df)
      - replace_zero_with_NaN_list: list of column names to replace zeros with NaNs
    - return_just_summer (default True): return seasonally-aggregated data just for
      the 6 month growing season ('summer'), or also for the winter 6 months?

    Returns: dataframe of seasonally-aggregated data
    """
    import pandas as pd
    import numpy as np

    # Turn off "Setting with copy" warning, which is returning a false positive
    pd.options.mode.chained_assignment = None  # default='warn'

    # Drop any dictionary keys that aren't needed
    for key in list(agg_method_dict.keys()):
        if key not in daily_df.columns:
            del agg_method_dict[key]

    # Resample ('Q' for quarterly, '-month' for month to end in). If season
    # function changes, need to change this too
    # Returned df: winter values are stored against the year that corresponds
    # to the second half of the winter (e.g. Nov 99-Apr 2000 stored as 2000).
    # The shift is needed because otherwise the last day of the period
    # (corresponding to the label) is omitted. Checked manually and right.
    season_df = daily_df.shift(periods=-1).resample('2Q-Apr', closed='left').agg(agg_method_dict)

    # np.nansum isn't working as I expected, with sums of 0 when all values
    # are NaN. Replace these with NaNs.
    for col in replace_zero_with_NaN_li:
        season_df.loc[season_df[col] == 0.0, col] = np.NaN

    # Remove frequncy info from index so plotting works right
    season_df.index.freq = None

    # Add season info to the dataframe, i.e. a column with either "summer"
    # or "winter"
    def season(x):
        """Input month number, and return the season it corresponds to
        """
        if x in [11, 12, 1, 2, 3, 4]:
            return 'wint'
        else:
            return 'summ'

    season_df['Season'] = season_df.index.month.map(season)

    if return_just_summer is True:
        # Remove winter rows (a bit long-winded, but works)
        summer_df = season_df.loc[season_df['Season'] == 'summ']
        summer_df.drop('Season', axis=1, inplace=True)

        # Reindex
        summer_df['year'] = summer_df.index.year
        summer_df.set_index('year', inplace=True)

        return summer_df

    else:
        return season_df


def discretize(thresholds, value):
    """
    Function to compare a number to a list of thresholds and categorise accordingly.
    E.g. to apply row-wise down a df to convert from continuous to categorical data.
    Input:
        thresholds: list of class boundaries that define classes of interest
        value: float to be compared to the thresholds

    Returns: class the value lies in. Classes are defined in factor_li_dict
    within function according to number of thresholds
    (max 2 class boundaries (thresholds) supported at present)

    e.g. of usage:
    # E.g. 1:
    bound_dict = {'TP':[29.5]}
    for col in continuous_df.columns:
        disc_df[col] = continuous_df[col].apply(lambda x: discretize(bound_dict['TP'], x))
    # E.g. 2:
    df['WFD_class'] = df[['threshold','expected_value']].apply(lambda x: discretize([x.threshold], x.expected_value), axis=1)
    """
    import numpy as np

    if np.isnan(value):
        return np.NaN

    factor_li_dict = {2: [0, 1],
                      3: [0, 1, 2]}

    n_classes = len(thresholds) + 1

    for i, boundary in enumerate(thresholds):

        if value < boundary:
            return factor_li_dict[n_classes][i]
            break  # Break out of loop

        # If we're up to the last class boundary, and the value is bigger than it,
        # value is in the uppermost class
        if i + 1 == len(thresholds) and value >= boundary:
            return factor_li_dict[n_classes][i+1]